server/services/v1/auth_provider.go

```diff
@@ -299,45 +326,159 @@ func getAccessTokenUsingRefreshToken(req *api.GetAccessTokenRequest, a *Auth0) (
+
+	defer func(Body io.ReadCloser) {
+		if err := Body.Close(); err != nil {
+			log.Warn().Err(err).Msg("could not close the response body")
+		}
+	}(resp.Body)
+
 	body, err := io.ReadAll(resp.Body)
 	if err != nil {
-		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = %s", err.Error())
+		msg := "Failed to get access token: reason =" + err.Error()
+		return nil, tokenError(msg, msg, err)
 	}
 	bodyStr := string(body)
 	if resp.StatusCode == 200 {
 		getAccessTokenResponse := api.GetAccessTokenResponse{}
 
 		err = json.Unmarshal([]byte(bodyStr), &getAccessTokenResponse)
 		if err != nil {
-			return nil, api.Errorf(api.Code_INTERNAL, "Failed to parse external response: reason = %s", err.Error())
+			msg := "Failed to parse external response: reason = " + err.Error()
+			return nil, tokenError(msg, msg, err)
+		}
+		// cache it
+		tx, err := a.txMgr.StartTx(ctx)
+		if err != nil {
+			return nil, tokenError("Failed to start tx to insert into cache", "Failed to get access token: reason = could not process cache", err)
+		}
+
+		cacheEntry := &tokenMetadataEntry{
+			AccessToken: getAccessTokenResponse.GetAccessToken(),
+			ExpireAt:    time.Now().Add(time.Second * time.Duration(getAccessTokenResponse.GetExpiresIn())).Unix(),
+		}
+		cacheEntryBytes, err := json.Marshal(&cacheEntry)
+		if err != nil {
+			_ = tx.Rollback(ctx)
+			return nil, tokenError("Failed to start marshal cache entry to json bytes", "Failed to get access token: reason = could not process cache", err)
+		}
+
+		err = a.userStore.InsertUserMetadata(ctx, tx, default_namespace_id, metadata.Application, req.GetClientId(), metadataKey, cacheEntryBytes)
+		if err != nil {
+			_ = tx.Rollback(ctx)
+			return nil, tokenError("Failed to insert entry into cache", "Failed to get access token: reason = could not process cache", err)
+		}
+
+		err = tx.Commit(ctx)
+		if err != nil {
+			return nil, tokenError("Failed to commit transaction to insert entry into cache", "Failed to get access token: reason = could not process cache", err)
 		}
+
 		return &getAccessTokenResponse, nil
 	}
 	log.Error().Msgf("Auth0 response status code=%d", resp.StatusCode)
 	return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = %s", bodyStr)
 }
 
+func tokenError(logMsg string, userfacingErrorMsg string, err error) error {
+	log.Warn().Err(err).Msg(logMsg)
+	return api.Errorf(api.Code_INTERNAL, userfacingErrorMsg)
+}
+
 func readDate(dateStr string) int64 {
 	result, err := date.ToUnixMilli(time.RFC3339, dateStr)
 	if err != nil {
-		log.Warn().Msgf("%s field was not parsed to int64", dateStr)
+		log.Warn().Err(err).Msgf("%s field was not parsed to int64", dateStr)
 		result = -1
 	}
 	return result
 }
+
+func createAccessTokenMetadataKey(clientSecret string) string {
+	hash := sha256.Sum256([]byte(clientSecret))
+	encodedHash := base64.StdEncoding.EncodeToString(hash[:])
+	return access_token_using_client_creds_prefix + encodedHash
```

Comment:
Should prefix be clientSecret instead of access_token...?

Reply:
we can keep access_token in the keyspace to differentiate it from other metadata that may come in future.